tests/unit_tests/dft/source/descriptor_tests.cpp

```diff
@@ -0,0 +1,453 @@
+
+        descriptor.get_value(oneapi::mkl::dft::config_param::FORWARD_SCALE,
+                             &forward_scale_before_set);
+        EXPECT_EQ(1.0, forward_scale_before_set);
+        descriptor.set_value(oneapi::mkl::dft::config_param::FORWARD_SCALE,
+                             forward_scale_set_value);
+        descriptor.get_value(oneapi::mkl::dft::config_param::FORWARD_SCALE,
+                             &forward_scale_after_set);
+        EXPECT_EQ(forward_scale_set_value, forward_scale_after_set);
+    }
+
+    {
+        Precision_Type backward_scale_set_value{ 143.5 };
+        Precision_Type backward_scale_before_set;
+        Precision_Type backward_scale_after_set;
+
+        descriptor.get_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE,
+                             &backward_scale_before_set);
+        EXPECT_EQ(1.0, backward_scale_before_set);
+        descriptor.set_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE,
+                             backward_scale_set_value);
+        descriptor.get_value(oneapi::mkl::dft::config_param::BACKWARD_SCALE,
+                             &backward_scale_after_set);
+        EXPECT_EQ(backward_scale_set_value, backward_scale_after_set);
+    }
+
+    {
+        std::int64_t n_transforms_set_value{ 12 };
+        std::int64_t n_transforms_before_set;
+        std::int64_t n_transforms_after_set;
+
+        descriptor.get_value(oneapi::mkl::dft::config_param::NUMBER_OF_TRANSFORMS,
+                             &n_transforms_before_set);
+        EXPECT_EQ(1, n_transforms_before_set);
+        descriptor.set_value(oneapi::mkl::dft::config_param::NUMBER_OF_TRANSFORMS,
+                             n_transforms_set_value);
+        descriptor.get_value(oneapi::mkl::dft::config_param::NUMBER_OF_TRANSFORMS,
+                             &n_transforms_after_set);
+        EXPECT_EQ(n_transforms_set_value, n_transforms_after_set);
+    }
+
+    {
+        std::int64_t fwd_distance_set_value{ 12 };
+        std::int64_t fwd_distance_before_set;
+        std::int64_t fwd_distance_after_set;
+
+        descriptor.get_value(oneapi::mkl::dft::config_param::FWD_DISTANCE,
+                             &fwd_distance_before_set);
+        EXPECT_EQ(1, fwd_distance_before_set);
+        descriptor.set_value(oneapi::mkl::dft::config_param::FWD_DISTANCE, fwd_distance_set_value);
+        descriptor.get_value(oneapi::mkl::dft::config_param::FWD_DISTANCE, &fwd_distance_after_set);
+        EXPECT_EQ(fwd_distance_set_value, fwd_distance_after_set);
+    }
+
+    {
+        std::int64_t bwd_distance_set_value{ 12 };
+        std::int64_t bwd_distance_before_set;
+        std::int64_t bwd_distance_after_set;
+
+        descriptor.get_value(oneapi::mkl::dft::config_param::BWD_DISTANCE,
+                             &bwd_distance_before_set);
+        EXPECT_EQ(1, bwd_distance_before_set);
+        descriptor.set_value(oneapi::mkl::dft::config_param::BWD_DISTANCE, bwd_distance_set_value);
```

Comment:
Setting this to be the same as the forward distance with COMPLEX_CONJUGATE (for real inplace problems), would result in a inconsistent configuration error as the bwd_distance, should be half the fwd_distance.

Reply:
We are testing real inplace transforms in a [next MR](https://github.com/oneapi-src/oneMKL/pull/275/files#diff-8bab6f36b34ac3bdbdeb4b7966c34be0dc994201f919ac42b63a37fd0cb313f2R116). This sets the same value for forward and backward distance and passes. The [spec](https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/dft/config_params/distance.html#onemkl-dft-config-distance) only says that the 2 distances should be consistent. Are you talking about another case that we are not testing yet?

Reply:
The fairly ambiguous "consistency" is a (hidden) reference to "For in-place transforms, ensure the first element of the input data has the same location as the first element of the output data for each dimension." from [this page of our docs](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/fourier-transform-functions/fft-functions/configuration-settings/dfti-complex-real-conj-even-storage.html). For in-place complex transforms, this means equal forward and backward distances, indeed. However, for real transform with `DFTI_CONJUGATE_EVEN_STORAGE=DFTI_COMPLEX_COMPLEX`, the (implicit) units used for strides and distances are real floating point values in forward domain and complex floating point values in backward domains (more details on the implicit "units" in either domain [here](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/fourier-transform-functions/fft-functions/configuration-settings/dfti-input-strides-dfti-output-strides.html)). So, the backward distance must be half the forward distance in that case (a padding of two real floating point values is required in the forward domain).

Reply:
Can you explain why the code @Rbiessy linked works then, even for cases with NUMBER_OF_TRANSFORMS > 1? Is BWD_DISTANCE simply ignored?

Reply:
I believe the test lines linked by @Rbiessy works becasue the container size is setup correctly, I am talking about this tests specifically `/DescriptorTestSuite/DescriptorTests.DescriptorTestsRealSingle/` we are setting up both the input and output distance as 12 (with a bunch of other parameters) and the eventual recipe is commited (line 344), that commit will fail, because of what @raphael-egan explanations.

Reply:
From what I understand the 2 tests have a similar configuration, only the one I linked is running the computation after committing the descriptor. I would expect the MKLGPU backend to throw or produce incorrect results in the test I linked if we set the wrong backward distance. Is it passing by chance? I don't understand what do you mean by "the container size is setup correctly"? I may need to have a closer look tomorrow.

Reply:
Ah! my bad, distance wise this is setup the same only difference I can find between these two is not setting up  
`descriptor.set_value(oneapi::mkl::dft::config_param::CONJUGATE_EVEN_STORAGE,
                             oneapi::mkl::dft::config_value::COMPLEX_COMPLEX);`
maby this reference can help with the configuration : [dfti_create_node.c#L259](https://github.com/intel-innersource/libraries.performance.math.mkl/blob/07b4a95a09cef51e7d6cd728fe6051ea91dcd9cc/src/dft/kernel/0ref/dfti_create_node.c#L259)

Reply:
In the open-source [spec](https://spec.oneapi.io/versions/latest/elements/oneMKL/source/domains/dft/config_params/storage_formats.html#conjugate-even-storage) `CONJUGATE_EVEN_STORAGE` only supports `COMPLEX_COMPLEX` so this is implicitly set [here](https://github.com/oneapi-src/oneMKL/pull/259/files#diff-5493672861f50e819aba7deb8468890d0d6e0d2d92cf600b89a9f412f825bbb8R109). This makes sense since there is a [note](https://www.intel.com/content/www/us/en/develop/documentation/onemkl-developer-reference-c/top/fourier-transform-functions/fft-functions/configuration-settings/dfti-complex-real-conj-even-storage.html#dfti-complex-real-conj-even-storage) saying `DFTI_CONJUGATE_EVEN_STORAGE=DFTI_COMPLEX_REAL` is deprecated (and will be removed for multi-dimensional FFTs).
We have noticed a couple of strange behaviors with the MKLGPU backend:
- I was not able to query the backend's fwd and bwd distance. It throws an exception with `FFT_INVALID_DESCRIPTOR` for all of our tests. Is that expected?
- It seems the backend ignores the value set for `BWD_DISTANCE` as our multi-batches test pass for any value we set. I can half the `BWD_DISTANCE` for real transforms but it seems there is backend's bug?

Reply:
- `FFT_INVALID_DESCRIPTOR` being returned when querying fwd/bwd distance used to be a bug on our side: [MKLD-13592](https://jira.devtools.intel.com/browse/MKLD-13592). It was fixed on 11/03/2022 for oneMKL2023.0 release I believe, in [PR#2163](https://github.com/intel-innersource/libraries.performance.math.mkl/pull/2163);
- regarding the `BWD_DISTANCE` value apparently ignored, I believe you are right although that depends on the problem configuration and the targeted GPU device (could you provide more details?). Unless you're testing an r2c of size 2*N <= 4096 on PVC, the default code path is taken and the bwd distance is indeed ignored for in-place operation in that case. It is instead coerced into the value that's required for consistency. I agree that it should at least check that the user did that as well and error out if not: this is part of something I have intended to address for quite a while now ([MKLD-13390](https://jira.devtools.intel.com/browse/MKLD-13390)) but not in our top priorities so far unfortunately

Reply:
Thank you for clarifying.

- I am using the 2023.0 release, maybe it will be fixed for the next release?
- We are testing some rather old integrated GPU (HD Graphics 530 in my case) and the configuration is real, in-place, single precision, size=2x2x2, batch=2, backward scale=1/8, fwd_distance = bwd_distance = 8. How can the backend force a bwd_distance? Isn't the user allowed to set a larger bwd_distance in such a case?

I have removed the commit of the descriptor for this particular to make it clearer that this is only testing the open-source descriptor (1961d9c0605e9dc8bd367b708afa65c1bdec14b2) here.
I have halved the distance for this descriptor even if that shouldn't matter anymore (2aba40dbfccb127331e2ba8211836f3c23a0d430).

Reply:
Sorry, I edited my previous message, the distances are set to 16 indeed. I'd like to discuss in our next meeting if we can clarify the situation for the users regarding bwd_distance. That conversation got a bit off topic!